zingstats/changes.py

```diff
@@ -0,0 +1,287 @@
+
+
+class Changes(object):
+    def __init__(self, url, query, projects, branches, start_dt, finish_dt,
+                 verify_https_requests=False):
+        self.url = url
+        self.query = query
+        self.projects = projects
+        self.branches = branches
+        self.start_dt = start_dt
+        self.finish_dt = finish_dt
+        self.query_start = 0
+        self.verify_https_requests = verify_https_requests
+        self.changes = dict()
+
+    def __iter__(self):
+        for long_id in self.changes:
+            yield self.changes[long_id]
+
+    def __len__(self):
+        return len(self.changes)
+
+    def add(self, change):
+        self.changes[change.long_id] = change
+
+    # TODO implement better__getitem__()
+    def __getitem__(self, key):
+        return self.changes[key]
+
+    @staticmethod
+    def pretty_json(json_data):
+        """Pretty print JSON data, usually for logging purposes."""
+        return json.dumps(json_data, sort_keys=True, indent=4,
+                          separators=(',', ': '))
+
+
+class GerritChanges(Changes):
+    def __init__(self, url, query, projects, branches, start_dt, finish_dt,
+                 verify_https_requests=False,
+                 query_size=100, max_changes=None):
+        super(GerritChanges, self).__init__(url, query, projects, branches,
+                                            start_dt, finish_dt,
+                                            verify_https_requests)
+        self.query_size = query_size
+        self.max_changes = max_changes
+
+    def gather(self):
+        log.info('Gathering gerrit changes from %s for projects: %s and branches: %s',
+                self.url,
+                ', '.join(sorted(self.projects)),
+                ', '.join(sorted(self.branches)))
+
+        results = [{'_more_changes': True}]
+        session = requests.Session()
+        while '_more_changes' in results[-1].keys() and \
+                results[-1]['_more_changes']:
+            log.debug('Querying %d changes starting at %d', self.query_size,
+                      self.query_start)
+            payload = {
+                'q': self.query,
+                # TODO make ALL_FILES gathering toggleable, it is expensive/slow
+                #'o': ['ALL_REVISIONS', 'MESSAGES', 'ALL_FILES'],
+                'o': ['ALL_REVISIONS', 'MESSAGES'],
+                'start': self.query_start,
+                'n': self.query_size}
+            query = ('%s/changes/' % self.url)
+            response = session.get(query, params=payload,
+                                   verify=self.verify_https_requests)
+            log.debug(response.url)
+
+            results = GerritChanges.clean_gerrit_response(response)
+            log.debug(GerritChanges.pretty_json(results))
+            for change_json in results:
+                log.debug(GerritChanges.pretty_json(change_json))
+                log.debug('%d changes (start: %d, count=%d)', len(results),
```

Comment:
Do you need to log this on each iteration?

Reply:
it's proven useful for debugging, maybe worth revisiting once I get better testing in place